### PR TITLE
xDS-failover: attempt reconnection to primary after disconnected from failover

### PR DESCRIPTION
**Pull request by adisuissa**
Commit Message: xDS-failover: attempt reconnection to primary after disconnected from failover
Additional Description:
Prior to this PR, Envoy would stick to either the primary or failover xDS source, once it got a response from one.
This PR changes the behavior by allowing Envoy to attempt reconnecting to the primary source, after being disconnected from the failover source.
Risk Level: low - xDS-failover is disabled by default.
Testing: Added unit and integration tests.
Docs Changes: N/A - docs will be created in a subsequent PR.
Release Notes: N/A - no release note yet as this is WIP.
Platform Specific Features: N/A

### CODE REVIEW

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h [hunk 1]
```diff
@@ -84,32 +87,43 @@ class GrpcMuxFailover : public GrpcStreamInterface<RequestType, ResponseType>,
...
+                     "Already connected to an xDS server, skipping establishNewStream() call");
+      return;
+    }
     // First check if Envoy ever connected to the primary/failover, and if so
     // persist attempts to that source.
-    if (ever_connected_to_ == ConnectedState::Primary) {
-      ASSERT(!connecting_to_failover_);
+    // Prior connection to failover is disregarded as Envoy will try
+    // reconnecting to the primary again.
+    if (ever_connected_to_primary_) {
```
Comment (nezdolik) [inline source/extensions/config_subscription/grpc/grpc_mux_failover.h:106]: why do we need `ever_connected_to_primary_` field? Would it not be enough just to try to connect to primary by default if `connection_state_ == ConnectionState::None || connection_state_ == ConnectionState::ConnectingToPrimary`?

**Review by nezdolik — COMMENTED** [review]
lgtm, but had one question

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h (hunk 1 above)
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h]: Good question.
The requirement is to be never connect to the failover source after Envoy was connected to the primary.
Note that the `ever_connected_to_primary_` is also used in [line 219](https://github.com/envoyproxy/envoy/blob/1f53139b64319a903170c3ba190ccc61bc4dec42/source/extensions/config_subscription/grpc/grpc_mux_failover.h#L219). There it is used as a marker to remember not to reconnect to the primary, so the flag is still needed.

In the context of this method, however, I agree that it is redundant, and having less code is probably better in this case. I've removed this piece of code, and added a comment explaining why the requirement holds. Thanks for pointing this out!

**Review by nezdolik — APPROVED** [review]

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h [hunk 2]
```diff
@@ -30,19 +30,23 @@ namespace Config {
...
- * will be used.
- * If the failover source is unavailable, the GrpcMuxFailover will alternate between attempts
- * to reconnect to the primary source and the failover source.
- * In the future, this behavior may change, and the GrpcMuxFailover will always
- * prefer the primary source, even if prior connection to the failover was successful.
+ * failover source. If a response from the failover source is received, it will be the source of
+ * configuration until the connection is closed. In case the connection is closed, Envoy will
+ * attempt to connect to the primary, before retrying to connect to the failover source.
+ * If the failover source is unavailable or a connection to it is closed, the GrpcMuxFailover
+ * will alternate between attempts to reconnect to the primary source and the failover source.
```
Comment (htuch) [inline source/extensions/config_subscription/grpc/grpc_mux_failover.h:49]: Is this true on startup a well?

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h [hunk 3]
```diff
@@ -191,21 +195,19 @@ class GrpcMuxFailover : public GrpcStreamInterface<RequestType, ResponseType>,
...
-      ASSERT(parent_.connectingToOrConnectedToPrimary() &&
-             !parent_.connectingToOrConnectedToFailover());
+      ASSERT(parent_.connectingToOrConnectedToPrimary());
       // If there's no failover supported, this will just be a pass-through
       // callback.
       if (parent_.failover_grpc_stream_ != nullptr) {
-        if (parent_.connecting_to_primary_ &&
-            (parent_.ever_connected_to_ != ConnectedState::Primary)) {
+        if (parent_.connection_state_ == ConnectionState::ConnectingToPrimary &&
+            !parent_.ever_connected_to_primary_) {
```
Comment (htuch) [inline source/extensions/config_subscription/grpc/grpc_mux_failover.h:210]: Is there an opportuntiy to scatter some more ASSERTs with this variable around?

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h [hunk 4]
```diff
@@ -84,32 +87,34 @@ class GrpcMuxFailover : public GrpcStreamInterface<RequestType, ResponseType>,
...
-      return;
-    } else if (ever_connected_to_ == ConnectedState::Failover) {
-      ASSERT(!connecting_to_primary_);
-      ENVOY_LOG_MISC(trace, "Attempting to reconnect to the failover gRPC source, as a connection "
-                            "to it was previously established.");
-      establishStreamToFailoverIfNotConnected();
+    // If already connected to one of the source, return.
+    if (connection_state_ == ConnectionState::ConnectedToPrimary ||
+        connection_state_ == ConnectionState::ConnectedToFailover) {
+      ENVOY_LOG_MISC(trace,
```
Comment (htuch) [inline source/extensions/config_subscription/grpc/grpc_mux_failover.h:98]: Missing test coverage.

**Review by htuch — COMMENTED** [review]
LGTM modulo comments.

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h (hunk 3 above)
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h:211]: Done

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h (hunk 2 above)
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h]: Yeah... I tried to rephrase this comment. Feel free to point out if it should be further improved.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**